### ISSUE TITLE
Remember last 'who' for faster re-ticking (was 'with' field autocomplete should have a new 'recently used' group list of users)

### ISSUE THREAD

**Issue by brendanheywood** [theCrag/website#2813]
**What happened?**

Especially on mobile but also for desktop, it's way quicker to add ticks, and importantly add ticks correctly if you can easily select the people you are climbing with without really typing anything. In almost all cases you are very likely to climb with someone who is the same person you climbed with recently. If we stored a recent climbed with list for each account you could select 'with' in 2 clicks and move on faster.

This is exacerbated when you are ticking one route at a time. A lot of power users, myself included, usually tick each route individually for various reasons. This means I don't get the benefit from the batch 'with' defaults and fill out the form over and over again. Even when I am using a batch I may have climbed different routes with different people and still want to log each with, and so I still get the same problem. Over the last week I've exclusively used my phone for ticking so this problem has jumped out at me along with the 2 previous autocomplete issues (#2811, #2812).

**What you expected:**

*Simpler method using last who*

* [ ] If I tick a single route, copy the 'who' field into a session var or somewhere semi persistent
* [ ] If I tick a group of ticks, copy the 'default with' field into the same session var as above
* [ ] If I go to tick a single route, then make the default whatever is in the session
* [ ] If I go to tick a batch, put the session var back into the 'default with'

This also has another big advantage over the idea below because it will work with names which are not accounts which is extremely common. Looking back at my tick history, roughly half of my ticks have a who who doesn't have an account. It's a lot of manual typing.

The last bit to think about is the duration that it honors the session var. 24 hours seems very safe, but it's very likely over a weekend you will climb with the same people. I think 48 hours would be a good threshold. 

*Second method using most common whos*

* [ ] store a small number of unique recent climbers for each account. I'd say 5 is a good number as it will almost but not quite fill the dropdown box, and as soon as you type anything then you'll either filter that list of get a completely new list of other people. Any higher and it means you'll have to scroll through the list. It could be smaller but I'd say 3 would be the minimum as most people would have a couple regularly partners.

* [ ] would we just pick the 5 most recently used unique climbers? this seems the simplest the implement and the most bang for buck.

**Comment by scd** [comment]
I don't think we have to store anything, just define a new text completion rule. It has been a while since I looked at the mentions internals, but if I am right, at the end of the line of minor config changes we will have to add a sql query.

**Comment by brendanheywood** [comment]
I'm not fussed if it's stored or dynamic, same same

**Comment by brendanheywood** [comment]
I've been mulling on this a bit more and I think we could do something even simpler which would be even better. For instance today I ticked about 15-20 boulders but spread over almost as many different boulders. So I have to go through the ticking process multiple times. Every time I'm typing in the same set of names over and over. Most of them are repeats so I'm just clicking stuff and not typing many comments, the main thing I'm typing is just the who. On some occasions in a big sessions in a big group I might log different 'whos' but in general I think it's a very safe bet that if I log more than once in the same day, then defaulting to the same people I previously ticked would make things a lot faster. Even in the event that some of the people have changed, removing people is faster than adding them so it less of a burden, and you might lose one from the party and gain another etc so still benefit from some common state to start from.

So this boils down to:
* [ ] If I tick a single route, copy the 'who' field into a session var or somewhere semi persistent
* [ ] If I tick a group of ticks, copy the 'default with' field into the same session var as above
* [ ] If I go to tick a single route, then make the default whatever is in the session
* [ ] If I go to tick a batch, put the session var back into the 'default with'

This also has another big advantage over the previous idea because it will work with names which are not accounts which is extremely common. Looking back at my tick history, roughly have of my ticks have a who who doesn't have an account. It's a lot of manual typing.

The last bit to think about is the duration that it honors the session var. 24 hours seems very safe, but it's very likely over a weekend you will climb with the same people. I think 48 hours would be a good threshold.

**Comment by brendanheywood** [comment]
This also came up in the context of gyms, if people are ticking per route, or per rope, this will make things much faster

**Comment by lordyavin** [comment]
+1

**Comment by brendanheywood** [comment]
+1 from support email

**Comment by rouletout** [comment]
+1 from support email to record "non-useres" as climbing partners

**Comment by lordyavin** [comment]
+1 from me

**Comment by lordyavin** [comment]
Waiting for that to be implemented. For every individual logbook entry I have to lookup my main climbing partner again. It's quite annoying.

BTW: I don't think wee need session info here. I simple query on the logbook would do the trick.

    SELECT DISTINCT with_field
    FROM user_logbook
    ORDER BY ascent_date DESC

**Comment by brendanheywood** [comment]
There are two use cases:

1) I climbed this week with the same person I climbed with last week (or usually)

2) I'm logging multiple ticks in a single day, but in different cliffs / boulders

The session method works better for use case 1. The 'recent climbers' or 'common partners' dropdown would improve use case 2.

As a boulderer I mostly tick something here and then something there on lots of different pages. A single session with 15 ticks might be 5 tick screens and I have to type the same things in every time. I might on average climb with the same group but it will be a different mix each time.

Another reasons similar to above, even when route climbing I tend to tick things in the order I did them not just batch tick. It would help if we could re-order routes in the ticking page, but that doesn't cover the case when I want to log working and sending attempts of the same route in the same session out of order.

If I am ticking in the same day, this should be auto filled and zero typing, it's not even a new autocomplete 'recent' section, which would also have value in other situations.
